src/Plutip/Types.purs

```diff
@@ -197,27 +216,138 @@ instance DecodeAeson StopClusterResponse where
 -- | distribution for wallets.
 -- | Number of wallets in distribution specification matches the number of
 -- | wallets provided to the user.
-class UtxoDistribution distr wallets | distr -> wallets, wallets -> distr where
+class UtxoDistribution distr wallets | distr -> wallets where
   encodeDistribution :: distr -> Array (Array UtxoAmount)
-  decodeWallets :: Array PrivateKeyResponse -> Maybe wallets
+  decodeWallets
+    :: PrivatePaymentKey
+    -> distr
+    -> Array PrivateKeyResponse
+    -> Contract () wallets
 
 instance UtxoDistribution Unit Unit where
   encodeDistribution _ = []
-  decodeWallets _ = Just unit
+  decodeWallets _ _ _ = pure unit
 
 instance UtxoDistribution InitialUTxO KeyWallet where
   encodeDistribution amounts = [ amounts ]
-  decodeWallets [ (PrivateKeyResponse key) ] =
+  decodeWallets _ _ [ PrivateKeyResponse key ] =
     pure $ privateKeysToKeyWallet (PrivatePaymentKey key) Nothing
-  decodeWallets _ = Nothing
+  decodeWallets _ _ _ = wrongNumberPrivateKeysError
 
 instance
   UtxoDistribution restSpec restWallets =>
   UtxoDistribution (InitialUTxO /\ restSpec) (KeyWallet /\ restWallets) where
-  encodeDistribution (amounts /\ rest) =
-    encodeDistribution amounts <> encodeDistribution rest
-  decodeWallets = Array.uncons >>> case _ of
-    Nothing -> Nothing
-    Just { head: PrivateKeyResponse key, tail } ->
-      Tuple (privateKeysToKeyWallet (PrivatePaymentKey key) Nothing) <$>
-        decodeWallets tail
+  encodeDistribution = encodeDistributionRecurse
+  decodeWallets ourKey (distr /\ rest) = decodeWalletsRecurse ourKey distr rest
+
+instance UtxoDistribution InitialUTxOWithStakeKey KeyWallet where
+  encodeDistribution (InitialUTxOWithStakeKey _ amounts) = [ amounts ]
+  decodeWallets ourKey (InitialUTxOWithStakeKey stake _) [ key ] =
+    decodeWallet ourKey key (Just stake)
+  decodeWallets _ _ _ = wrongNumberPrivateKeysError
+
+instance
+  UtxoDistribution restSpec restWallets =>
+  UtxoDistribution (InitialUTxOWithStakeKey /\ restSpec)
```

Comment:
Shouldn't we generalize these two recursive instances?

```
  UtxoDistribution headSpec headWallet =>
  UtxoDistribution restSpec restWallets =>
  UtxoDistribution (headSpec /\ restSpec) (headWallet /\ restWallets) where
```

and remove `encodeDistributionRecurse`